### PR TITLE
CORE-6497 Uniqueness checker message bus integration

### PR DESCRIPTION
**Pull request by relyafi**
### Summary

This change exposes the uniqueness checker component to the message bus. This is achieved by introducing a durable subscription and corresponding durable processor that listens to messages on the new uniqueness check topic introduced in the corresponding API PR (https://github.com/corda/corda-api/pull/575). The message process simply delegates to the existing request processing logic in the uniqueness checker. Responses are returned using the external event response API. This allows a notarisation flow to request uniqueness checking and receive responses using the external events API.

### Testing

A new set of integration tests, `MessageBusIntegrationTests` has been introduced, which uses the DB based message bus to ensure that messages can be sent to the uniqueness check topic, processed correctly by the uniqueness checker, and an appropriate response is placed on the flow events topic for consumption by the external events implementation. These tests are not exhaustive, and complement the existing tests:

- `UniquenessCheckerImplTests`: Unit tests that directly test `BatchedUniquenessCheckerImpl`, using a fake backing store
- `UniquenessCheckerImplDBIntegrationTests`: Integration tests that directly test `BatchedUniquenessCheckerImpl`, using `JPABackingStore` with a real database (HSQLDB or Postgres)
- `MessageBusIntegrationTests`: Integration tests that place requests on the DB based message bus and listen to responses on the flow events topic. Processing performed by `BatchedUniquenessCheckerImpl`, using a fake backing store

The new tests are a small subset of the tests used in the unit tests; the intent is simply to ensure that we correctly receive each type of success and error response back from the flow events topic. Comprehensive functional testing is carried out by the existing tests.

Also manually spun up a corda cluster and verified that the DB worker continues to start successfully.

### CODE REVIEW

**Comment by relyafi** [comment]
@lankydan included you specifically as thought you might be interested in [TestExternalEventResponseMonitor.kt](https://github.com/corda/corda-runtime-os/pull/2119/files#diff-b5e95c82bfa892e9231c15c13949b3a9de67856f9686aefbac280fa42d49d321)

File: components/uniqueness/uniqueness-checker-client-service-impl/src/main/kotlin/net/corda/uniqueness/client/impl/UniquenessCheckerClientServiceImpl.kt [hunk 1]
```diff
@@ -65,7 +68,14 @@ class UniquenessCheckerClientServiceImpl @Activate constructor(
     ): UniquenessCheckResponse {
         log.info("Received request with id: $txId, sending it to Uniqueness Checker")
 
+        @Suppress("ForbiddenComment")
+        // TODO: CORE-4730 to pass through the Vnode holding id plus a sensible event context
```
Comment (relyafi) [inline components/uniqueness/uniqueness-checker-client-service-impl/src/main/kotlin/net/corda/uniqueness/client/impl/UniquenessCheckerClientServiceImpl.kt:72]: @nkovacsx this is the bit that will need revisiting as part of your flow worker integration ticket

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerImpl.kt [hunk 2]
```diff
@@ -59,12 +73,29 @@ class BatchedUniquenessCheckerImpl(
     constructor(
         @Reference(service = LifecycleCoordinatorFactory::class)
         coordinatorFactory: LifecycleCoordinatorFactory,
+        @Reference(service = ConfigurationReadService::class)
+        configurationReadService: ConfigurationReadService,
+        @Reference
```
Comment (nkovacsx) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerImpl.kt]: Is the `service` property intentionally missing here?

File: testing/flow/external-events/src/main/kotlin/net/corda/test/flow/externalevents/TestExternalEventResponseMonitor.kt [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import net.corda.messaging.api.subscription.config.SubscriptionConfig
+import net.corda.messaging.api.subscription.factory.SubscriptionFactory
+import net.corda.schema.Schemas
+import net.corda.test.util.eventually
+import java.time.Duration
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.TimeoutException
+
+/**
+ * Helper class for listening to responses to requests that use the external events API. This can be
```
Comment (nkovacsx) [inline testing/flow/external-events/src/main/kotlin/net/corda/test/flow/externalevents/TestExternalEventResponseMonitor.kt]: Seems like there's a typo here

File: components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt [hunk 4]
```diff
@@ -0,0 +1,424 @@
...
+import org.osgi.test.junit5.service.ServiceExtension
+import java.time.Duration
+import java.time.Instant
+import java.time.LocalDate
+import java.time.ZoneOffset
+import java.util.*
+
+/**
+ * Tests the integration of the uniqueness checker component with the message bus. This duplicates
+ * a small subset of the tests in [UniquenessCheckerImplTests][net.corda.uniqueness.checker.impl.UniquenessCheckerImplTests]
```
Comment (nkovacsx) [inline components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt]: Line length is inconsistent here

File: components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+                        ConfigKeys.MESSAGING_CONFIG,
+                        Configuration(
+                            MESSAGING_CONFIG,
+                            MESSAGING_CONFIG,
+                            0,
+                            ConfigurationSchemaVersion(1,0)
+                        )
+                    )
+                )
+            )[0]
```
Comment (nkovacsx) [inline components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt]: I'm not a fan of [] accessor in Kotlin, could we use `first`, `single`?

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ */
+class UniquenessCheckMessageProcessor(
+    private val uniquenessChecker: UniquenessChecker,
+    private val externalEventResponseFactory: ExternalEventResponseFactory
+): DurableProcessor<String, UniquenessCheckRequestAvro> {
+
+    override val keyClass = String::class.java
+    override val valueClass = UniquenessCheckRequestAvro::class.java
+
+    override fun onNext(events: List<Record<String, UniquenessCheckRequestAvro>>): List<Record<*, *>> {
```
Comment (nkovacsx) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt:21]: Is there a reason we use this exact return type? Will that be used anywhere?

File: testing/flow/external-events/src/main/kotlin/net/corda/test/flow/externalevents/TestExternalEventResponseMonitor.kt (hunk 3 above)
Comment (nkovacsx) [reply testing/flow/external-events/src/main/kotlin/net/corda/test/flow/externalevents/TestExternalEventResponseMonitor.kt]: My bad, disregard this comment

File: components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt (hunk 5 above)
Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt]: I think we can in this case, but more generally, just FYI I think what you're saying goes against convention - e.g. if you use get(0) then you get told by IntelliJ that you should use indexing instead.

Comment (nkovacsx) [reply components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt]: Sorry I probably didn't phrase it right. I meant if we want the first element of the list, I'd rather use `first` or `firstOrNull` than `[0]`. Obviously if we want a different element (not the first or last) then `[]` is the Kotlin way. :)

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt (hunk 6 above)
Comment (nkovacsx) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt:21]: Ah I see it's coming from the base class, nevermind!

File: components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt (hunk 5 above)
Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl-osgi-tests/src/integrationTest/kotlin/net/corda/uniqueness/checker/impl/osgitests/MessageBusIntegrationTests.kt]: Yep that makes sense - changed to use single as we are expecting exactly one result.

File: testing/flow/external-events/src/main/kotlin/net/corda/test/flow/externalevents/TestExternalEventResponseMonitor.kt [hunk 7]
```diff
@@ -0,0 +1,85 @@
+package net.corda.test.flow.externalevents
```
Comment (lankydan) [inline testing/flow/external-events/src/main/kotlin/net/corda/test/flow/externalevents/TestExternalEventResponseMonitor.kt]: would prefer `external.events` package to follow the same naming of the actual library.

File: testing/flow/external-events/src/main/kotlin/net/corda/test/flow/externalevents/TestExternalEventResponseMonitor.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                    currentData.values.forEach { processEvent(it) }
+
+                override fun onNext(
+                    newRecord: Record<String, FlowEvent>,
+                    oldValue: FlowEvent?,
+                    currentData: Map<String, FlowEvent>
+                ) = processEvent(newRecord.value)
+
+                private fun processEvent(event: FlowEvent?) {
+                    if (event != null && event.payload is ExternalEventResponse) {
```
Comment (lankydan) [inline testing/flow/external-events/src/main/kotlin/net/corda/test/flow/externalevents/TestExternalEventResponseMonitor.kt]: Should be able to do `(event.payload as? ExternalEventResponse)?.let { eeResponse ->`

**Review by nkovacsx — DISMISSED** [review]
Looks good now!

File: components/uniqueness/uniqueness-checker-impl-osgi-tests/test.bndrun [hunk 9]
```diff
@@ -0,0 +1,46 @@
+-tester: biz.aQute.tester.junit-platform
+-runfw: org.apache.felix.framework
+-resolve.effective: resolve,active
```
Comment (lankydan) [inline components/uniqueness/uniqueness-checker-impl-osgi-tests/test.bndrun:3]: Do the OSGi tests have to be a separate module?

Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl-osgi-tests/test.bndrun:3]: Technically I don't think they have to be, but I wanted to keep them separate from the existing non-OSGi integration tests we already have which have a lot more dependencies, as I spent a lot of time dealing with OSGi dependency problems, so keeping it compartmentalised keeps it simple.

It looks like this pattern has already been used elsewhere in the code base - e.g. `libs/db` similarly has an `osgi-integration-tests` module. Some others achieve a similar thing by using a separate gradle task for non-OSGi tests. I felt the separate module approach was a bit cleaner.

**Review by lankydan — DISMISSED** [review]

**Review by ronanbrowne — DISMISSED** [review]
LGTM ( Build changes)

**Review by nkovacsx — DISMISSED** [review]

**Review by nkovacsx — DISMISSED** [review]

**Review by ronanbrowne — APPROVED** [review]
LGTM (Build)

**Pull request merged**